### DISCUSSION TITLE
Can we stop the page load or redirect after clicking a button?

### DISCUSSION THREAD

**Discussion by page-source** [cypress-io/cypress#16064] [Questions and Help]
I have a button in my App which when clicked redirects to another page. My requirement while testing the click of the button is that I don't want the app to redirect to the next page. Instead just be on the page where it is. And this I am asking because when the button is clicked, the window object gets updated with data around click and I need to test that data in window object. Problem is when the redirect happens, the window object doesn't contain the properties from previous page, of course. Do we have something where we can ignore the redirect or ignore the XHR requests?
My code is something as below -
```
it('Button Click Event Fired',() => {
  cy.visit(env+'/');
  const btnLabel = "get_started";
  cy.get('button.cta').trigger("click"); // click happens here, window object gets updated as soon as btn is clicked
  cy.window().then(windowObj => {. // need to test the window object without redirecting to other page
    return checkAnalytics(btnLabel, windowObj)
  }).should('eq', true)
});
```

**Comment by fmoliveira** [comment]
Hi Satyam!

First of all, I wouldn't recommend trying to stop or ignore the redirect, because the redirect seems to be an actual part of the user flow. Disrupting the user flow to assert something will have you testing for something different from what real users will experience, which could hide real bugs from your plain sight.

Let me check if I understood your flow:
- user presses the button
- app sends out an XHR request
- XHR response is assigned to a window object
- your test want to assert the value of that window object assigned by the XHR response?

If that's it, did you consider using [cy.intercept](https://docs.cypress.io/api/commands/intercept)? With that you'll be able to spy on the XHR request and block your test flow until the XHR response is received (set an alias to your intercepted request, and wait for that alias). This part may be tricky if your redirect is triggered too fast, but this can potentially be the right time to check if the desired data is there. In short, you'd try to validate the data before the redirect, and just let the redirect of the user flow continue naturally.

Anyway, in the event that you really need to intervene with the redirect, that depends on how your redirect is triggered. Is that with the `history` package, directly with `window.location`, with a form submit? Stubbing the redirect from `window.location` may be doable although I still haven't tried that, but in essence, you'll need to figure out what triggers the redirect and try to stub that. If you do this, do not make it a regular practice and reserve it for last-resort situations. Testing real user interactions will significantly improve the quality of your e2e tests and simplify their maintenance.

PS: also consider looking for the received XHR response in the form of something displayed on the page, if possible, instead of looking up directly on the implementation details under the window object
